fluture.js

```diff
@@ -376,544 +404,927 @@
+  }
+
+  FutureTry.prototype.toString = function FutureTry$toString(){
+    return `Future.try(${show(this._fn)})`;
+  }
+
+  //----------
+
+  function FutureEncase(fn, a, b, c){
+    this._length = arguments.length - 1;
+    this._fn = fn;
+    this._a = a;
+    this._b = b;
+    this._c = c;
+    this._f = FutureEncase.FS[this._length];
+  }
+
+  FutureEncase.FS = {
+    1: function FutureEncase$1$fork(rej, res){
       let r;
-      try{ r = f(x) }catch(e){ return void rej(e) }
+      try{ r = this._fn(this._a) }catch(e){ rej(e); return noop }
       res(r);
       return noop;
-    });
-  };
-
-  Future.encase2 = function Future$encase2(f, x, y){
-    check$encase2(f);
-    if(arguments.length === 1) return unaryPartial(Future.encase2, f);
-    if(arguments.length === 2) return binaryPartial(Future.encase2, f, x);
-    return new FutureClass(false, function Future$encase2$fork(rej, res){
+    },
+    2: function FutureEncase$2$fork(rej, res){
       let r;
-      try{ r = f(x, y) }catch(e){ return void rej(e) }
+      try{ r = this._fn(this._a, this._b) }catch(e){ rej(e); return noop }
       res(r);
       return noop;
-    });
-  };
-
-  Future.encase3 = function Future$encase3(f, x, y, z){
-    check$encase3(f);
-    if(arguments.length === 1) return unaryPartial(Future.encase3, f);
-    if(arguments.length === 2) return binaryPartial(Future.encase3, f, x);
-    if(arguments.length === 3) return ternaryPartial(Future.encase3, f, x, y);
-    return new FutureClass(false, function Future$encase3$fork(rej, res){
+    },
+    3: function FutureEncase$3$fork(rej, res){
       let r;
-      try{ r = f(x, y, z) }catch(e){ return void rej(e) }
+      try{ r = this._fn(this._a, this._b, this._c) }catch(e){ rej(e); return noop }
       res(r);
       return noop;
+    }
+  }
+
+  FutureEncase.prototype = Object.create(Future.prototype);
+
+  FutureEncase.prototype.toString = function FutureEncase$toString(){
+    const args = [this._a, this._b, this._c].slice(0, this._length).map(show).join(', ');
+    const name = `encase${this._length > 1 ? this._length : ''}`;
+    return `Future.${name}(${show(this._fn)}, ${args})`;
+  }
+
+  //----------
+
+  function FutureChain(parent, chainer){
+    this._parent = parent;
+    this._chainer = chainer;
+  }
+
+  FutureChain.prototype = Object.create(Future.prototype);
+
+  FutureChain.prototype._f = function FutureChain$fork(rej, res){
+    const _this = this;
+    let cancel;
+    const r = _this._parent._f(rej, function FutureChain$fork$res(x){
+      const m = _this._chainer(x);
+      check$chain$f(m, _this._chainer, x);
+      cancel = m._f(rej, res);
     });
-  };
+    return cancel ? cancel : (cancel = r, function FutureChain$fork$cancel(){ cancel() });
+  }
 
-  Future.try = function Future$try(f){
-    return Future.encase(f, undefined);
-  };
+  FutureChain.prototype.toString = function FutureChain$toString(){
+    return `${this._parent.toString()}.chain(${showf(this._chainer)})`;
```

Comment:
My first reaction to this convention was it's nice to see the parent string stringified but now I'm wondering about long Future chains and what usefulness that monster string would provide.